src/ToDoContext.tsx

```diff
@@ -0,0 +1,164 @@
+/* eslint-disable no-param-reassign */
+// eslint-disable-next-line
+import _ from 'lodash';
+import React, { useEffect, useReducer } from 'react';
+import { ToDo } from './types/ToDo';
+
+type Action = { type: 'addPost', payload: string }
+| { type: 'updatePost', payload: ToDo }
+| { type: 'removePost', payload: ToDo }
+| { type: 'completed', payload: number }
+| { type: 'sortBy', payload: string }
+| { type: 'TOGGLE_ALL', payload: boolean }
+| { type: 'removeComplited' };
+
+interface State {
+  list: ToDo[];
+  sortBy: string;
+}
+
+function newToDo(name: string): ToDo {
+  return { id: Date.now(), title: name, completed: false };
+}
+
+function toggle(list: ToDo[], id: number): ToDo[] {
+  const copyList = _.cloneDeep(list);
+
+  copyList.forEach(todo => {
+    if (todo.id === id) {
+      todo.completed = !todo.completed;
+    }
+  });
+
+  return copyList;
+}
+
+function toggleAllHelper(elem: ToDo, trigger: boolean) {
+  if (!trigger) {
+    elem.completed = true;
+  } else {
+    elem.completed = false;
+  }
+
+  return elem;
+}
+
+function update(list: ToDo[], toDo: ToDo): ToDo[] {
+  const copyList = _.cloneDeep(list);
+  const index = copyList.findIndex(todo => todo.id === toDo.id);
+
+  copyList.splice(index, 1, toDo);
+
+  return copyList;
+}
+
+function remove(list: ToDo[], toDo: ToDo): ToDo[] {
+  const copyList = _.cloneDeep(list);
+  const index = copyList.findIndex(todo => todo.id === toDo.id);
+
+  copyList.splice(index, 1);
+
+  return copyList;
+}
+
+function reducer(state: State, action: Action): State {
+  switch (action.type) {
+    case 'addPost':
+      return {
+        ...state,
+        list: [...state.list, newToDo(action.payload)],
+      };
+    case 'updatePost':
+      return {
+        ...state,
+        list: update(state.list, action.payload),
+      };
+    case 'removePost':
+      return {
+        ...state,
+        list: remove(state.list, action.payload),
+      };
+    case 'completed':
+      return {
+        ...state,
+        list: toggle(state.list, action.payload),
+      };
+    case 'sortBy':
+      return {
+        ...state,
+        sortBy: action.payload,
+      };
+    case 'removeComplited':
+      return {
+        ...state,
+        list: [...state.list.filter(elem => !elem.completed)],
+      };
+    case 'TOGGLE_ALL':
+      return {
+        ...state,
+        list: [...state.list.map(elem => toggleAllHelper(elem, action.payload)),
+        ],
+      };
+    default:
+      return state;
+  }
+}
+
+const localStorageData = () => {
+  const data = localStorage.getItem('list');
+  let parsedData: ToDo[] = [];
+
+  if (data !== null) {
+    parsedData = JSON.parse(data);
+  } else {
+    return parsedData;
+  }
+
+  return parsedData;
+};
+
+const initialState: State = {
+  list: localStorageData(),
+  sortBy: 'ALL',
+};
+
+export const StateContext = React.createContext(initialState);
+/* eslint-disable */
+export const DispatchContext = React.createContext((_action: Action) => { });
```

Comment:
```suggestion
export const DispatchContext = React.createContext(() => {});
```

Reply:
unfortunately it throw the error about a lack of arguments
<img width="1280" alt="Screenshot 2023-08-23 at 15 01 17" src="https://github.com/mate-academy/react_todo-app/assets/20277989/40ce9b5c-50fb-4fae-bd24-a8f9320cb336">
<img width="1280" alt="Screenshot 2023-08-23 at 15 01 25" src="https://github.com/mate-academy/react_todo-app/assets/20277989/14a56905-14bf-45f7-945e-05292cb16b34">
<img width="1280" alt="Screenshot 2023-08-23 at 15 01 31" src="https://github.com/mate-academy/react_todo-app/assets/20277989/d40c4db3-3cc1-42bf-8b2d-8708c67cd0db">